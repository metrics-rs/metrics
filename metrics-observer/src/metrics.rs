use std::collections::HashMap;
use std::io::Read;
use std::net::TcpStream;
use std::net::ToSocketAddrs;
use std::sync::{Arc, Mutex, RwLock};
use std::thread;
use std::time::Duration;

use bytes::{BufMut, BytesMut};
use hdrhistogram::Histogram;
use prost::Message;

use metrics::{KeyData, Label, Unit};
use metrics_util::{CompositeKey, MetricKind};

mod proto {
    include!(concat!(env!("OUT_DIR"), "/event.proto.rs"));
}

use self::proto::{
    event::Event,
    metadata::{Description as DescriptionMetadata, MetricType, Unit as UnitMetadata},
    Event as EventWrapper,
};

#[derive(Clone)]
pub enum ClientState {
    Disconnected(Option<String>),
    Connected,
}

#[derive(Clone)]
pub enum MetricData {
    Counter(u64),
    Gauge(f64),
    Histogram(Histogram<u64>),
}

pub struct Client {
    state: Arc<Mutex<ClientState>>,
    metrics: Arc<RwLock<HashMap<CompositeKey, MetricData>>>,
    metadata: Arc<RwLock<HashMap<(MetricKind, String), (Option<Unit>, Option<String>)>>>,
    handle: thread::JoinHandle<()>,
}

impl Client {
    pub fn new(addr: String) -> Client {
        let state = Arc::new(Mutex::new(ClientState::Disconnected(None)));
        let metrics = Arc::new(RwLock::new(HashMap::new()));
        let metadata = Arc::new(RwLock::new(HashMap::new()));
        let handle = {
            let state = state.clone();
            let metrics = metrics.clone();
            let metadata = metadata.clone();
            thread::spawn(move || {
                let mut runner = Runner::new(addr, state, metrics, metadata);
                runner.run();
            })
        };

        Client {
            state,
            metrics,
            metadata,
            handle,
        }
    }

    pub fn state(&self) -> ClientState {
        self.state.lock().unwrap().clone()
    }

    pub fn get_metrics(&self) -> Vec<(CompositeKey, MetricData, Option<Unit>, Option<String>)> {
        let metrics = self.metrics.read().unwrap();
        let metadata = self.metadata.read().unwrap();

        metrics
            .iter()
            .map(|(k, v)| {
                let metakey = (k.kind(), k.key().name().to_string());
                let (unit, desc) = match metadata.get(&metakey) {
                    Some((unit, desc)) => (unit.clone(), desc.clone()),
                    None => (None, None),
                };

                (k.clone(), v.clone(), unit, desc)
            })
            .collect()
    }
}

enum RunnerState {
    Disconnected,
    ErrorBackoff(&'static str, Duration),
    Connected(TcpStream),
}

struct Runner {
    state: RunnerState,
    addr: String,
    client_state: Arc<Mutex<ClientState>>,
    metrics: Arc<RwLock<HashMap<CompositeKey, MetricData>>>,
    metadata: Arc<RwLock<HashMap<(MetricKind, String), (Option<Unit>, Option<String>)>>>,
}

impl Runner {
    pub fn new(
        addr: String,
        state: Arc<Mutex<ClientState>>,
        metrics: Arc<RwLock<HashMap<CompositeKey, MetricData>>>,
        metadata: Arc<RwLock<HashMap<(MetricKind, String), (Option<Unit>, Option<String>)>>>,
    ) -> Runner {
        Runner {
            state: RunnerState::Disconnected,
            addr,
            client_state: state,
            metrics,
            metadata,
        }
    }

    pub fn run(&mut self) {
        loop {
            let next = match self.state {
                RunnerState::Disconnected => {
                    // Just reset the client state here to be sure.
                    {
                        let mut state = self.client_state.lock().unwrap();
                        *state = ClientState::Disconnected(None);
                    }

                    // Try to connect to our target and transition into Connected.
                    let addr = match self.addr.to_socket_addrs() {
                        Ok(mut addrs) => match addrs.next() {
                            Some(addr) => addr,
                            None => {
                                let mut state = self.client_state.lock().unwrap();
                                *state = ClientState::Disconnected(Some(
                                    "failed to resolve specified host".to_string(),
                                ));
                                break;
                            }
                        },
                        Err(_) => {
                            let mut state = self.client_state.lock().unwrap();
                            *state = ClientState::Disconnected(Some(
                                "failed to resolve specified host".to_string(),
                            ));
                            break;
                        }
                    };
                    match TcpStream::connect_timeout(&addr, Duration::from_secs(3)) {
                        Ok(stream) => RunnerState::Connected(stream),
                        Err(_) => RunnerState::ErrorBackoff(
                            "error while connecting",
                            Duration::from_secs(3),
                        ),
                    }
                }
                RunnerState::ErrorBackoff(msg, dur) => {
                    {
                        let mut state = self.client_state.lock().unwrap();
                        *state = ClientState::Disconnected(Some(format!(
                            "{}, retrying in {} seconds...",
                            msg,
                            dur.as_secs()
                        )));
                    }
                    thread::sleep(dur);
                    RunnerState::Disconnected
                }
                RunnerState::Connected(ref mut stream) => {
                    {
                        let mut state = self.client_state.lock().unwrap();
                        *state = ClientState::Connected;
                    }

                    let mut buf = BytesMut::new();
                    let mut rbuf = [0u8; 1024];

                    loop {
                        match stream.read(&mut rbuf[..]) {
                            Ok(0) => break,
                            Ok(n) => buf.put_slice(&rbuf[..n]),
                            Err(e) => eprintln!("read error: {:?}", e),
                        };

                        let event = match EventWrapper::decode_length_delimited(&mut buf) {
                            Err(e) => {
                                eprintln!("decode error: {:?}", e);
                                continue;
                            }
                            Ok(event) => event,
                        };

                        if let Some(event) = event.event {
                            match event {
                                Event::Metadata(metadata) => {
                                    let metric_type = MetricType::from_i32(metadata.metric_type)
                                        .expect("unknown metric type over wire");
                                    let metric_type = match metric_type {
                                        MetricType::Counter => MetricKind::Counter,
                                        MetricType::Gauge => MetricKind::Gauge,
                                        MetricType::Histogram => MetricKind::Histogram,
                                    };
                                    let key = (metric_type, metadata.name);
                                    let mut mmap = self
                                        .metadata
                                        .write()
                                        .expect("failed to get metadata write lock");
                                    let entry = mmap.entry(key).or_insert((None, None));
                                    let (uentry, dentry) = entry;
                                    *uentry = metadata
                                        .unit
                                        .map(|u| match u {
                                            UnitMetadata::UnitValue(us) => us,
                                        })
                                        .and_then(|s| Unit::from_str(s.as_str()));
                                    *dentry = metadata.description.map(|d| match d {
                                        DescriptionMetadata::DescriptionValue(ds) => ds,
                                    });
                                }
                                Event::Metric(metric) => {
                                    let mut labels_raw =
                                        metric.labels.into_iter().collect::<Vec<_>>();
                                    labels_raw.sort_by(|a, b| a.0.cmp(&b.0));
                                    let labels = labels_raw
                                        .into_iter()
                                        .map(|(k, v)| Label::new(k, v))
                                        .collect::<Vec<_>>();
                                    let key_data: KeyData = (metric.name, labels).into();

                                    match metric.value.expect("no metric value") {
                                        proto::metric::Value::Counter(value) => {
                                            let key = CompositeKey::new(
                                                MetricKind::Counter,
                                                key_data.into(),
                                            );
                                            let mut metrics = self.metrics.write().unwrap();
                                            let counter = metrics
                                                .entry(key)
                                                .or_insert_with(|| MetricData::Counter(0));
                                            if let MetricData::Counter(inner) = counter {
                                                *inner += value.value;
                                            }
                                        }
                                        proto::metric::Value::Gauge(value) => {
                                            let key = CompositeKey::new(
                                                MetricKind::Gauge,
                                                key_data.into(),
                                            );
                                            let mut metrics = self.metrics.write().unwrap();
                                            let gauge = metrics
                                                .entry(key)
                                                .or_insert_with(|| MetricData::Gauge(0.0));
                                            if let MetricData::Gauge(inner) = gauge {
                                                *inner = value.value;
                                            }
                                        }
                                        proto::metric::Value::Histogram(value) => {
                                            let key = CompositeKey::new(
                                                MetricKind::Histogram,
                                                key_data.into(),
                                            );
                                            let mut metrics = self.metrics.write().unwrap();
                                            let histogram =
                                                metrics.entry(key).or_insert_with(|| {
                                                    let histogram = Histogram::new(3)
                                                        .expect("failed to create histogram");
                                                    MetricData::Histogram(histogram)
                                                });

                                            if let MetricData::Histogram(inner) = histogram {
                                                inner
                                                    .record(value.value)
                                                    .expect("failed to record value to histogram");
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }

                    RunnerState::ErrorBackoff("error while observing", Duration::from_secs(3))
                }
            };
            self.state = next;
        }
    }
}
