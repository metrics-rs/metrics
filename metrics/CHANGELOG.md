# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- next-header -->

## [Unreleased] - ReleaseDate

## [0.14.2] - 2021-02-13
### Added
- Implemented `Ord`/`PartialOrd` for various key-related types.

## [0.14.1] - 2021-02-02
### Added
- Minor documentation test updates for better coverage of owned strings used as metric names.

## [0.14.0] - 2021-02-02
### Changed
- Added support for owned strings as metric names. [#170](https://github.com/metrics-rs/metrics/pull/170)

## [0.13.1] - 2021-01-23
### Added
- Added conversion from `std::borrow::Cow<'static, str>` for `SharedString`.

## [0.13.0] - 2021-01-22
### Added
- New macros for registration: `register_counter!`, `register_gauge!`, `register_histogram!`.
- New macros for emission: `histogram!`, `increment_counter!`, `increment_gauge!`,
  `decrement_gauge!`.
- Added unit support to describe the unit of a given metric.

### Removed
- Dropped the `timing!` and `value!` macros in favor of `histogram!`.

### Changed
- All macros are now procedural macros instead of declarative macros.
- Gauges are now `f64` instead of `i64`.
- Histograms are now `f64` instead of `u64`.
- `Key` now split into `Key` and `KeyData` to better support constant expression metric keys via
  macro callsites.
- Significant overhaul of generated callsites via macros.

## [0.12.1] - 2019-11-21
### Changed
- Cost for macros dropped to almost zero when no recorder is installed. ([#55](https://github.com/metrics-rs/metrics/pull/55))

## [0.12.0] - 2019-10-18
### Changed
- Improved documentation. (#44, #45, #46)
- Renamed `Recorder::record_counter` to `increment_counter` and `Recorder::record_gauge` to `update_gauge`. ([#47](https://github.com/metrics-rs/metrics/pull/47))

## [0.11.1] - 2019-08-09
### Changed
- Fixed a bug with macros calling inner macros without a fully qualified name.

## [0.11.0] - 2019-07-29
### Added
- Life begins at 0.11.0 for this crate, after being renamed from `metrics-facade` to `metrics` to
  reflect the duality of `metrics` to the `log` crate. (#27)
