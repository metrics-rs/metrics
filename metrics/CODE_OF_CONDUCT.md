# The Code of Conduct

This document is based on the [Rust Code of Conduct](https://www.rust-lang.org/conduct.html) and outlines the standard of conduct which is both expected and enforced as part of this project.

## Conduct

**Contact**: [toby@nuclearfurnace.com](mailto:toby@nuclearfurnace.com)

* We are committed to providing a friendly, safe and welcoming environment for all, regardless of level of experience, gender identity and expression, sexual orientation, disability, personal appearance, body size, race, ethnicity, age, religion, nationality, or other similar characteristic.
* Avoid using overtly sexual nicknames or other nicknames that might detract from a friendly, safe and welcoming environment for all.
* Please be kind and courteous. There's no need to be mean or rude.
* Respect that people have differences of opinion and that every design or implementation choice carries a trade-off and numerous costs. There is seldom a right answer.
* Please keep unstructured critique to a minimum. If you have solid ideas you want to experiment with, make a fork and see how it works.
* We will exclude you from interaction if you insult, demean or harass anyone. That is not welcome behaviour. We interpret the term "harassment" as including the definition in the <a href="http://citizencodeofconduct.org/">Citizen Code of Conduct</a>; if you have any lack of clarity about what might be included in that concept, please read their definition. In particular, we don't tolerate behavior that excludes people in socially marginalized groups.
* Private harassment is also unacceptable. No matter who you are, if you feel you have been or are being harassed or made uncomfortable by a community member, please contact one of the repository Owners immediately. Whether you're a regular contributor or a newcomer, we care about making this community a safe place for you and we've got your back.
* Likewise any spamming, trolling, flaming, baiting or other attention-stealing behaviour is not welcome.

## Moderation

These are the policies for upholding our community's standards of conduct. If you feel that a thread needs moderation, please use the contact information above, or mention @tobz in the thread.

1. Remarks that violate this Code of Conduct, including hateful, hurtful, oppressive, or exclusionary remarks, are not allowed. (Cursing is allowed, but never targeting another user, and never in a hateful manner.)
2. Remarks that moderators find inappropriate, whether listed in the code of conduct or not, are also not allowed.

In the Rust community we strive to go the extra step to look out for each other. Don't just aim to be technically unimpeachable, try to be your best self. In particular, avoid flirting with offensive or sensitive issues, particularly if they're off-topic; this all too often leads to unnecessary fights, hurt feelings, and damaged trust; worse, it can drive people away from the community entirely.

And if someone takes issue with something you said or did, resist the urge to be defensive. Just stop doing what it was they complained about and apologize. Even if you feel you were misinterpreted or unfairly accused, chances are good there was something you could've communicated better â€” remember that it's your responsibility to make your fellow Rustaceans comfortable. Everyone wants to get along and we are all here first and foremost because we want to talk about cool technology. You will find that people will be eager to assume good intent and forgive as long as you earn their trust.
