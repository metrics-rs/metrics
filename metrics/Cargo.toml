[package]
name = "metrics"
version = "0.9.0"
authors = ["Toby Lawrence <toby@metrics-rs.com>"]
edition = "2018"

license = "MIT"

description = "high-speed metrics collection library"

homepage = "https://github.com/metrics-rs/metrics"
repository = "https://github.com/metrics-rs/metrics"
documentation = "https://docs.rs/metrics"

readme = "README.md"

keywords = ["metrics", "telemetry", "histogram", "counter", "gauge"]

[profile.release]
debug = true
opt-level = 3
lto = true

[dependencies]
metrics-core = "^0.1"
crossbeam-channel = "^0.3"
parking_lot = "^0.7"
fnv = "^1.0"
hashbrown = "^0.1"
quanta = { version = "^0.2", features = ["asm"] }
derivative = "^1.0"
tokio-sync = "^0.1"

[dev-dependencies]
log = "^0.4"
env_logger = "^0.6"
getopts = "^0.2"
hdrhistogram = "^6.1"
