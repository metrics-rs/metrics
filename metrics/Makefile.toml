[env]
RUST_TEST_THREADS = "1"
CARGO_MAKE_RUN_CODECOV = "true"
RUSTFLAGS = "-D warnings"
CARGO_FEATURES = "--no-default-features"

[tasks.format-stable]
command = "cargo"
args = ["fmt"]

[tasks.pre-verify-project]
dependencies = ["check-format", "clippy", "coverage-flow"]

[tasks.pre-coverage]
dependencies = ["test"]

[tasks.build-verbose]
description = "Runs the rust compiler with verbose output, with interpolated features."
category = "Build"
command = "cargo"
args = ["build", "--verbose", "${CARGO_FEATURES}"]

[tasks.test-verbose]
description = "Runs all available tests with verbose output, with interpolated features."
category = "Test"
command = "cargo"
args = ["test", "--verbose", "${CARGO_FEATURES}"]

[tasks.test]
description = "Runs all available tests, with interpolated features."
category = "Test"
command = "cargo"
args = ["test", "${CARGO_FEATURES}"]
