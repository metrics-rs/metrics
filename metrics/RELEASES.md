# Releases
Unlike the [CHANGELOG](CHANGELOG.md), this file tracks more complicated changes that required
long-form description and would be too verbose for the changelog alone.

<!-- next-header -->

## [Unreleased] - ReleaseDate

- No notable changes.

## [0.20.1] - 2022-07-22

- No notable changes.

## [0.20.0] - 2022-07-20

- No notable changes.

## [0.19.0] - 2022-05-30

- No notable changes.

## [0.18.1] - 2022-03-10

- No notable changes.

## [0.18.0] - 2022-01-14

### Switch to metric handles
`metrics` has now switched to "metric handles."  This requires a small amount of backstory.

#### Evolution of data storage in `metrics`

Originally, [`metrics`][metrics] started its life as [`metrics-runtime`][metrics-runtime], which represented a batteries-included
implementation of metrics storage, with both a user-accessible API that allowed directly interacting
with the metrics as well as a mode for being installed as the global recorder for [`metrics`].

The user-accessible API was most commonly accessed via a helper type called `Sink`, which could both
create metric type-specific handles -- such as `Counter` or `Gauge` -- and could also perform
operations directly, such as `increment_counter`.

However, users still needed to both create the machinery supplied by `metrics-runtime` to handle
metrics storage, as well as the actual exporters to access those metrics, or send them to the
downstream system of choice.  We ultimately made the decision to bundle the metrics storage
capabilities into exporters directly by moving much of [`metrics-runtime`][metrics-runtime] to
[`metrics-util`][metrics-util] as a set of reusable components.

In doing so, though, we lost the ability to access metrics directly, as the [`metrics`][metrics] facade became
the only way to register or emit metrics, and this could only happen through the opaque macros,
going through the opaque `Recorder` trait, and so on.  This made it easy for all users to simply
interact with the macros, but made it harder to optimize the performance of recording metrics.

With this update, we've switched back to a handle-based model that also works seamlessly with the
macros.

#### Handle-based design

The `metrics` crate now provides concrete handle types for all metrics.  The `Recorder` API has also
been updated to return these handles when registering a metric.  Thus, for macros, instead of simply
calling the operation directly on the global recorder, they now register the metric first and call
the operation on the handle.  If the metric has already been registered, similar to as before, a
copy of the handle is simply returned and can be operated as normally.

On the flipside, though, users can use directly acquire these handles using the `register_*` macros.
The macros for adding descriptive information have been renamed to `describe_*`.  Thus, when there
is an opportunity to register metrics and store their handles somewhere for later reuse, users can
use the `register_*` macros to do so.

#### Implementation of handles

In order to provide an mechanism that allows enough flexibility for exporter authors, the design of the handle types was considered in depth.

Simply put, handles represent an incredibly thin layer over compatible implementations of each
metric type.  There are now three traits, one for each metric type, that must be implemented to
satisfy being wrapped in the corresponding handle type.  Internally, this is wrapped in an [`Arc<T>`][arc] to make it usable in multi-threaded scenarios.

Implementations of `Counter` and `Gauge` based on atomic integers can be found in
[`metrics`][metrics], and an example of `Histogram` can be found in [`metrics-util`][metrics-util].

#### Limitations of handles

In most cases, the new handle-based design will be effectively invisible, with no visible changes to
behavior.  There is one scenario where changes could be encountered, though: use of [layers][layers].

Layers work on a per-metric basis, which means that when a handle is acquired and used directly, any
layers that were used no longer have a way to intercept the calls made to handles.  When the macros
are used, however, a call is always made to "register" the metric, which provides a chance for the
layers to intercept and modify things as necessary.

As a concrete example, the [`metrics-tracing-context`][mtc] layer exists to annotate a metric's set of
labels with fields from the `tracing::Span` that is currently entered when the metric is updated.
Naturally, a given metric could be emitted while in a various number of different spans, but if a
metric is first registered in a particular span, and then the handle is used from that point on...
only the original span fields will be attached to the metric.

This is an inherent trade-off between providing a way to reduce the overhead of updating a metric to
the bare minimum, and providing extensibility via layers.

[metrics]: https://docs.rs/metrics/latest/metrics/
[metrics-runtime]: https://docs.rs/metrics-runtime/latest/metrics_runtime/
[metrics-util]: https://docs.rs/metrics-util/latest/metrics-util/
[arc]: https://doc.rust-lang.org/stable/std/sync/struct.Arc.html
[layers]: https://docs.rs/metrics-util/latest/metrics_util/layers/index.html
[mtc]: https://docs.rs/metrics-tracing-context/latest/metrics_tracing_context/
