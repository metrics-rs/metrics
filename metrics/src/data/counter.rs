use fnv::FnvBuildHasher;
use hashbrown::HashMap;
use crate::data::ScopedKey;

pub(crate) struct Counter {
    data: HashMap<ScopedKey, u64, FnvBuildHasher>,
}

impl Counter {
    pub fn new() -> Counter {
        Counter {
            data: HashMap::<ScopedKey, u64, FnvBuildHasher>::default(),
        }
    }

    pub fn update(&mut self, key: ScopedKey, delta: u64) {
        let value = self.data.entry(key).or_insert(0);
        *value = value.wrapping_add(delta);
    }

    pub fn values(&self) -> Vec<(ScopedKey, u64)> { self.data.iter().map(|(k, v)| (k.clone(), *v)).collect() }
}

#[cfg(test)]
mod tests {
    use super::{Counter, ScopedKey};

    #[test]
    fn test_counter_simple_update() {
        let mut counter = Counter::new();

        let key = ScopedKey(0, "foo".into());
        counter.update(key, 42);

        let key2 = ScopedKey(0, "foo".to_owned().into());
        counter.update(key2, 31);

        let values = counter.values();
        assert_eq!(values.len(), 1);
        assert_eq!(values[0].1, 73);
    }
}
