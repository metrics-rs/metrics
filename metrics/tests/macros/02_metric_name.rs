use metrics::counter;

fn valid_name() {
    counter!("abc_def", 1);
}

fn invalid_name() {
    counter!("abc$def");
}

fn main() {}
