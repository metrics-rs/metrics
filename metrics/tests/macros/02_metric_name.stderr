error: metric name must match ^[a-zA-Z][a-zA-Z0-9_:.]*$
 --> $DIR/02_metric_name.rs:8:14
  |
8 |     counter!("abc$def");
  |              ^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
